lib/_tls_wrap.js

```diff
@@ -370,7 +374,8 @@ function TLSSocket(socket, opts) {
   // Read on next tick so the caller has a chance to setup listeners
   process.nextTick(initRead, this, socket);
 }
-util.inherits(TLSSocket, net.Socket);
+Object.setPrototypeOf(TLSSocket.prototype, net.Socket.prototype);
+Object.setPrototypeOf(TLSSocket, net.Socket);
```

Comment:
Doesn't the second line undo what the first line already does? Why modify TLSSocket.prototype, and then replace it in the next line? I don't object if this is the new consensus way to implement inheritance, but the point of `util.inherits()` was more clear.

Reply:
@sam-github Can you explain this in more detail? The second line doesn’t undo the first, both `setPrototypeOf()` calls are independent (just to be clear, `a.setPrototypeOf(b)` does not modify any of the properties of `a`, so e.g. `a.prototype` still has its original value).

Reply:
We are replacing a lib function with clear purpose with calls to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf  that are not clear. I read the docs:

> The Object.setPrototypeOf() method sets the prototype (i.e., the internal [[Prototype]] property) of a specified object to another object

but it seems our `.prototype` is not the `[[prototype]]`.

> I don't object if this is the new consensus way to implement inheritance, but the point of util.inherits() was more clear.

I still don't object. I just grepped, I can see this is the way the code is going.

I also still don't understand how the code works, but I'll remember its purpose when I see it.

Reply:
> We are replacing a lib function with clear purpose with calls to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf that are not clear. I read the docs:
> 
> > The Object.setPrototypeOf() method sets the prototype (i.e., the internal [[Prototype]] property) of a specified object to another object
> 
> but it seems our `.prototype` is not the `[[prototype]]`.
> 

Yes, that’s correct. (And very unfortunate naming, in a way, but it’s built into the language).

> > I don't object if this is the new consensus way to implement inheritance, but the point of util.inherits() was more clear.
> 
> I still don't object. I just grepped, I can see this is the way the code is going.
> 
> I also still don't understand how the code works, but I'll remember its purpose when I see it.

The difference is that `util.inherits()` only sets the prototype of `ctor.prototype`, but not of `ctor` itself, but ES6 classes do (for practical purposes that means that static methods, not just instance methods, are inherited). And given that ES6 classes are “the” way to do inheritance in modern JS, it seems like we should align with that, even if we don’t use static methods a lot.